plugins/ReactTestComponent.js

```diff
@@ -0,0 +1,61 @@
+var reactTestInstance = Symbol.for('react.test.json');
+
+function handleChildren(result, children, indentation, print, indent) {
+  if (children) {
+    children.forEach(function(child) {
+      result.push('\n');
+      result.push(objectToJSX(child, indentation, print, indent));
+    });
+    result.push('\n');
+  }
+}
+
+function handleProps(result, node, indentation, print, indent) {
+  var props = node.props;
+  if (props) {
+    var indentOpts = {
+      indent: 2 * (indentation + 1)
+    };
+    Object.keys(props).forEach(function(prop) {
+      result.push('\n', indent(prop, indentOpts), '=');
+      var value = props[prop];
+      if (typeof value === 'string') {
+        result.push('"', value, '"');
+      } else {
+        var formatted = indent(print(value), {indent: 2 * (indentation + 2)});
+        result.push('{\n', formatted, '\n', indent('}', indentOpts));
+      }
+    });
+  }
+}
+
+function objectToJSX(root, indentation, print, indent) {
+  indentation = indentation || 0;
+  var indentationOpts = {indent: indentation * 2};
+  var type = root.type;
+  if (!type && typeof root === 'string'){
+    return indent(root, indentationOpts);
+  }
+  var result = [];
+  if (!root.children) {
+    result.push(indent('<', indentationOpts), type);
+    handleProps(result, root, indentation, print, indent);
+    result.push(' />');
+  } else {
+    result.push(indent('<', indentationOpts), type);
+    handleProps(result, root, indentation, print, indent);
+    result.push('>');
+    handleChildren(result, root.children, indentation + 1, print, indent);
+    result.push(indent('</', indentationOpts), type, '>');
+  }
+  return result.join('');
```

Comment:
While I think V8 optimise string concatenations by internally using an Array I' m not sure it does that for multiple statements, this can potentially use a lot of memory, and it should be especially true since I'm passing the result array to the handleProps/handleChildren so they can push immediately instead of building the result string and then returning it, (having to keep all the intermediate string in memory at the same time when handling the lowest leaf plus the result instead of having a simple linear array).